src/UI/Implementation/Component/Button/Renderer.php

```diff
@@ -122,6 +124,72 @@ protected function renderClose($component) {
 		return $tpl->get();
 	}
 
+	protected function renderToggle(Component\Button\Toggle $component) {
+		$tpl = $this->getTemplate("tpl.toggle.html", true, true);
+
+		$on_action = $component->getAction();
+		$off_action = $component->getActionOff();
+
+
+		if (is_string($on_action))
+		{
+			$on_url = $on_action;
+			$on_options = "{}";
+		}
+		else
+		{
+			$on_url = "";
+			$on_signal = $on_action->getSignal();
+			$on_event = $on_action->getEvent();
+			$on_options = json_encode($on_signal->getOptions());
+		}
+
+		if (is_string($off_action))
+		{
+			$off_url = $off_action;
+			$off_options = "{}";
+		}
+		else
+		{
+			$off_url = "";
+			$off_signal = $off_action->getSignal();
+			$off_event = $off_action->getEvent();
+			$off_options = json_encode($off_signal->getOptions());
+		}
+
+
+		if ($component->isActive()) {
+			$component = $component->withAdditionalOnLoadCode(function ($id)
+				use ($on_url, $on_signal, $on_event, $on_options,
+					$off_url, $off_signal, $off_event, $off_options) {
+				return "$('#$id').on('click', function(event) {
+						il.UI.button.handleToggleClick(event, '$id', '$on_url', {
+								'id' : '{$on_signal}', 'event' : '{$on_event}',
+								'triggerer' : $('#{$id}'),
+								'options' : JSON.parse('{$on_options}')}, '$off_url', {
+								'id' : '{$off_signal}', 'event' : '{$off_event}',
+								'triggerer' : $('#{$id}'),
+								'options' : JSON.parse('{$off_options}')
+							});
+						return false; // stop event propagation
+				});";
```

Comment:
I guess this will be much easier to understand if you'd construct the array on PHP-side and then use `json_decode `to pass it to js. This will spare the excessive use of `use` and hard to read array construction.

Reply:
I would go even somewhat further and factor this out to the js completely and just use some init function here. Like seen in [this](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/src/UI/Implementation/Component/Dropzone/File/Renderer.php#L144) example. However, you may decide if you really want to factor this out completely. However, please use json_decode, as @klees suggests.

Reply:
The structure reflects the signature of the jQuery.trigger function as being called in the AbstractComponentRenderer, too (addTriggererOnLoadCode). Of course it would be possible to completely encode it, but then I would need to additionally decode this on the JS side to get the parameters for trigger.